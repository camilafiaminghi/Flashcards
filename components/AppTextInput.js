import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, View, Text, TextInput } from 'react-native'
import { validationRules } from '../utils/validation'
import { gray, black } from '../utils/colors'

class AppTextInput extends Component {

	static propTypes = {
		name: PropTypes.string.isRequired,
		placeholder: PropTypes.string.isRequired,
		maxLen: PropTypes.number.isRequired,
		onInputChange: PropTypes.func
	}

	state = {
		text: '',
		changed: false,
		valid: false
	}

	handleValidation = (value) => {
		const { name, onInputChange, maxLen } = this.props
		const valid = validationRules(name, value)

		if ( value.length <= maxLen ) {
			this.setState((state) => ({
				...state,
				text: value,
				changed: true,
				valid
			}))

			onInputChange(name, value, valid)
		}
	}

	render() {
		const { text, changed, valid } = this.state
		const { maxLen, placeholder } = this.props
		const textLeft = maxLen - text.length

		return (
			<View style={styles.container}>
				<TextInput
	        style={[styles.input, styles.inputText]}
	        placeholder={placeholder}
	        onChangeText={this.handleValidation}
	        value={text} />

	      { (textLeft !== 0 && textLeft < maxLen && changed) && <Text style={styles.text}>{ textLeft } characteres left</Text> }
	      { (textLeft === 0 && changed) && <Text style={styles.text}>max characters exceeded</Text>}
			</View>
		)
	}
}

export default AppTextInput

/* STYLES */
const styles = StyleSheet.create({
	container: {
		alignSelf: 'stretch',
		margin: 10
	},
	input: {
		flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderColor: gray,
		borderWidth: 1,
		borderRadius: 4
	},
	inputText: {
		padding: 8
	},
	text: {
		fontSize: 12,
		color: black
	}
})
