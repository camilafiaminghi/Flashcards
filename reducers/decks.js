import {
		ADD_DECK,
		ADD_CARD } from '../actions'

const decks = (state = {}, action) => {
	switch (action.type) {
		case ADD_DECK:
			return {
				...state,
				[action.payload.title] : {
					title: action.payload.title,
					questions: []
				}
			}
		case ADD_CARD:
			return {
				...state,
				[action.payload.deckId] : {
					...state[action.payload.deckId],
					questions: [...state[action.payload.deckId].questions, action.payload.card]
				}
			}
		default:
			return state
	}
}

export default decks
