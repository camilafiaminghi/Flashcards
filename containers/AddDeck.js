import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { NavigationActions } from 'react-navigation'
import { connect } from 'react-redux'
import { StyleSheet, Text, View, KeyboardAvoidingView, TextInput, TouchableOpacity } from 'react-native'
import { handleAddDeck } from '../actions'
import { isValid } from '../utils/validation'
import { textColorInverse, textColor, pColor, pColorLight, pColorDark, sColor, sColorLight, sColorDark } from '../utils/colors'
import AppTextInput from '../components/AppTextInput'

class AddDeck extends Component {

	static propTypes = {
		onAddDeck: PropTypes.func.isRequired
	}

	state = {
		title: '',
		valid: false,
		resetValue: false
	}

	toDeck = () => {
		const { title } = this.state

		this.props.navigation.dispatch(NavigationActions.navigate({
			routeName: 'Deck',
			params: { deckId: title }
		}))
	}

	submit = () => {
		const { onAddDeck } = this.props
		const { title } = this.state

		onAddDeck({ title })
			.then(this.setState((state) => ({...state, resetValue: true})))
			.then(this.toDeck) /* Navigate to Decks */
	}

	onInputChange = (name, value, valid) => {
		this.setState((state) => ({
			...state,
			title: value,
			valid
		}))
	}

	componentDidMount() {
    this.didFocusListener = this.props.navigation.addListener(
      'didFocus',
      () => { this.setState((state) => ({...state, resetValue: false})) },
    )
  }

  componentWillUnmount() {
    this.didFocusListener.remove();
  }

	render() {
		const { valid, resetValue } = this.state

		return (
			<View style={styles.container}>
				<KeyboardAvoidingView behavior='padding' style={styles.card}>

					<Text style={styles.text}>What the title of your new deck?</Text>

					{ (!resetValue) &&
						<AppTextInput
							name='title'
							placeholder='Deck Title (max 55 characters)'
							maxLen={55}
							onInputChange={this.onInputChange}
							resetValue={resetValue} />
					}

				  <TouchableOpacity
		      	style={[styles.input, styles.btn]}
		      	disabled={!valid}
		      	onPress={this.submit}>
		      	<Text style={ valid ? styles.btnText : styles.btnDisabledText }>{'Create Deck'.toUpperCase()}</Text>
		      </TouchableOpacity>

				</KeyboardAvoidingView>
			</View>
		)
	}
}

export const mapDispatchToProps = (dispatch) => {
	return {
		onAddDeck: (deck) => dispatch(handleAddDeck(deck))
	}
}

export default connect(null, mapDispatchToProps)(AddDeck)

/* STYLES */
const styles = StyleSheet.create({
	container: {
		flex: 1,
    backgroundColor: sColorLight,
    paddingTop: 4,
    paddingBottom: 4
	},
	card: {
		flex: 1,
		flexGrow: 1,
		alignItems: 'center',
		justifyContent: 'center',
		backgroundColor: 'white',
		marginTop: 4,
  	marginBottom: 4,
  	marginRight: 8,
  	marginLeft: 8,
  	borderRadius: 3,
  	padding: 20
	},
	text: {
		textAlign: 'center',
  	fontSize: 24,
		marginRight: 24,
		marginLeft: 24,
		color: textColor
	},
	input: {
		flexDirection: 'row',
		alignSelf: 'stretch',
		height: 50,
		borderBottomColor: pColorDark,
		borderBottomWidth: 1,
		borderRadius: 3,
	},
	btn: {
		alignItems: 'center',
		justifyContent: 'center',
		padding: 4,
		marginRight: 10,
		marginLeft: 10
	},
	btnDisabledText: {
		fontSize: 14,
		fontWeight: 'bold',
		color: pColorLight
	},
	btnText: {
		fontSize: 14,
		fontWeight: 'bold',
		color: pColorDark
	}
})
