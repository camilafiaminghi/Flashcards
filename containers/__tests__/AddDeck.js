import React, { Component } from 'react'
import { TouchableOpacity } from 'react-native'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { AddDeck, mapDispatchToProps } from '../AddDeck'
import AppTextInput from '../../components/AppTextInput'
import decks from '../../__mocks__/decks'

let wrapper
let instance
let component
const navigation = {
	navigate: jest.fn(),
	addListener: jest.fn(),
	dispatch: jest.fn().mockResolvedValue()
}
const props = {
	navigation,
	onAddDeck: jest.fn().mockResolvedValue()
}

const configStore = configureMockStore([thunk])
const store = configStore({decks})

describe('<AddDeck />', () => {

	beforeEach(() => {
		wrapper = renderer.create(<AddDeck {...props} />)
		instance = shallow(<Provider store={store}><AddDeck {...props} /></Provider>)
		component = instance.find(AddDeck).shallow()
	})

	afterEach(() => store.clearActions())

	it('matches snapshot', () => {
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it('should render', () => {
		expect(wrapper).toBeTruthy()
	})

	it('should has one AppTextInput as child', () => {
		expect(component.find(AppTextInput)).toHaveLength(1)
	})

	it('should has submit if input is valid', () => {
		component.instance().onInputChange('title', 'Title Test', true)

		expect(component.state('valid')).toEqual(true)
		expect(component.find(TouchableOpacity)).toHaveLength(1)
	})

	it('should handle submit', () => {
		component.instance().onInputChange('title', 'Title Test', true)

		const spy = jest.spyOn(component.instance(), 'submit')
		const button = component.find(TouchableOpacity)

		button.props().onPress()
		expect(props.onAddDeck).toHaveBeenCalledTimes(1)
		expect(props.onAddDeck).toHaveBeenCalledWith({title: 'Title Test'})
	})

	it('should mapDispatchToProps return props', () => {
		expect(mapDispatchToProps(store.dispatch)).toHaveProperty('onAddDeck')
	})
})
