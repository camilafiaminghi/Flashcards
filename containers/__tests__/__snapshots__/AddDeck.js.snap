// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddDeck /> matches snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#efefef",
      "flex": 1,
      "paddingBottom": 4,
      "paddingTop": 4,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 3,
          "flex": 1,
          "flexGrow": 1,
          "justifyContent": "center",
          "marginBottom": 4,
          "marginLeft": 8,
          "marginRight": 8,
          "marginTop": 4,
          "padding": 20,
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#333333",
          "fontSize": 24,
          "marginLeft": 24,
          "marginRight": 24,
          "textAlign": "center",
        }
      }
    >
      What the title of your new deck?
    </Text>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "margin": 10,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Deck Title (max 55 characters)"
        placeholderTextColor="#6f79a8"
        style={
          Array [
            Object {
              "alignSelf": "stretch",
              "borderBottomColor": "#6f79a8",
              "borderBottomWidth": 2,
              "borderRadius": 3,
              "flexDirection": "row",
              "height": 50,
            },
            Object {
              "padding": 8,
            },
          ]
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "stretch",
          "borderBottomColor": "#6f79a8",
          "borderBottomWidth": 1,
          "borderRadius": 3,
          "flexDirection": "row",
          "height": 50,
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "padding": 4,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#d1d9ff",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        CREATE DECK
      </Text>
    </View>
  </View>
</View>
`;
